### PR TITLE
DAOS-11465 dfs: fix dfs_open for directories without O_EXCL

### PR DESCRIPTION
**Pull request by mchaarawi**
dfs_open with O_CREAT for directories always assume O_EXCL. While posix
does not provide a way to support that, we can easily support that in DFS
to have many processes/threads succeed when creating the same directory
with O_CREAT with only 1 creator and the rest just opening.

Signed-off-by: Mohamad Chaarawi <mohamad.chaarawi@intel.com>

### CODE REVIEW

File: src/client/dfs/dfs.c [hunk 1]
```diff
@@ -1192,51 +1226,69 @@ open_dir(dfs_t *dfs, dfs_obj_t *parent, int flags, daos_oclass_id_t cid,
...
-	 * O_CREATE and a entry of a different type exists already.
-	 */
+dopen:
+	/* Check that the opened object is the type that's expected. */
 	if (!S_ISDIR(entry->mode))
 		return ENOTDIR;
 
+	daos_mode = get_daos_obj_mode(flags);
+	if (daos_mode == -1)
+		return EINVAL;
```
Comment (daltonbohning) [inline src/client/dfs/dfs.c:1264]: Should we not use `D_GOTO(out, rc);` for consistency?

**Review by daltonbohning — DISMISSED** [review]

File: src/client/dfs/dfs.c (hunk 1 above)
Comment (mchaarawi) [reply src/client/dfs/dfs.c:1264]: yea good catch. need to fix that

**Review by daltonbohning — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

File: src/tests/suite/dfs_par_test.c [hunk 2]
```diff
@@ -739,75 +776,122 @@ file_atomicity_test_helper(test_arg_t *arg, int rf)
 		assert_non_null(oids_lo);
 	}
 
-	par_gather(PAR_COMM_WORLD, &oid1.hi, oids_hi, 1, PAR_UINT64, 0);
-	par_gather(PAR_COMM_WORLD, &oid1.lo, oids_lo, 1, PAR_UINT64, 0);
+	par_gather(PAR_COMM_WORLD, &foid1.hi, oids_hi, 1, PAR_UINT64, 0);
```
Comment (jolivier23) [inline src/tests/suite/dfs_par_test.c:779]: Not your issue here but couldn't you do this in a single gather with 2 entries instead of 1?

Comment (mchaarawi) [reply src/tests/suite/dfs_par_test.c:779]: yea for sure. a lot of optimizations can be made like this in the parallel tests.
but haven't given this much time to cleanup.

File: src/client/dfs/dfs.c [hunk 3]
```diff
@@ -1163,80 +1165,130 @@ open_dir(dfs_t *dfs, dfs_obj_t *parent, int flags, daos_oclass_id_t cid,
...
 
-	/* Check that the opened object is the type that's expected, this could
-	 * happen for example if dfs_open() is called with S_IFDIR but without
-	 * O_CREATE and a entry of a different type exists already.
-	 */
+dopen:
+	/* Check that the opened object is the type that's expected. */
 	if (!S_ISDIR(entry->mode))
-		return ENOTDIR;
+		D_GOTO(out, rc = ENOTDIR);
```
Comment (ashleypittman) [inline src/client/dfs/dfs.c:1260]: Is this correct?  rc will be set on L1287 by the check_tx call.

**Review by ashleypittman — CHANGES_REQUESTED** [review]

File: src/client/dfs/dfs.c (hunk 3 above)
Comment (ashleypittman) [reply src/client/dfs/dfs.c:1260]: Ah sorry, I forgot check_tx() would return the rc you passed it.

**Review by ashleypittman — APPROVED** [review]

**Pull request merged**